ykllvmwrap/src/jitmodbuilder.cc

```diff
@@ -137,11 +141,7 @@ class JITModBuilder {
       if (isRecursiveCall(CF)) {
         if (VMap[CF] == nullptr) {
           StringRef CFName = CF->getName();
-          // Declare function.
-          auto DeclFunc = llvm::Function::Create(CF->getFunctionType(),
-                                                 GlobalValue::ExternalLinkage,
-                                                 CFName, JITMod);
-          VMap[CF] = DeclFunc;
+          declareFunction(CF);
```

Comment:
Possibly silly question: can the same function be declared on both lines 121 and 144?

Reply:
If you declare a duplicate function name, LLVM will add a de-duplications suffix to the symbol name.

That said, I think `declareFunction()` should check so as to not bloat up the IR. It should probably also `assert` that the signature is the same?

Reply:
> That said, I think declareFunction() should check so as to not bloat up the IR. It should probably also assert that the signature is the same?

Agreed on both points. [Although -- if it's possible -- better than checking if the function has been declared would be to work out how not to double declare it in the first place?]

Reply:
It's inevitable when the trace calls the same symbol a second time.

Reply:
Ah!

Reply:
We already guard against that with `if (VMap[CF] == nullptr) {`. We never call `declareFunction` if it's already in the `VMap`.

Reply:
Even better. Still, an `assert` to make this clear in `declareFunction` might be a good idea.

Reply:
I think d730754 should do the trick. See if you agree at @vext01.

Reply:
LGTM